heaps/min_heap.py

```diff
@@ -57,18 +70,42 @@ def heap_up(self, index):
             property is reestablished.
             
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(log n)
+            Space complexity: O(log n)
         """
-        pass
+        if index == 0:
+            return
+        if self.store[index].key < self.store[(index - 1) // 2].key:
+            self.swap(index, (index - 1) // 2)
+            self.heap_up((index - 1) // 2)
 
     def heap_down(self, index):
         """ This helper method takes an index and 
             moves the corresponding element down the heap if it's 
             larger than either of its children and continues until
             the heap property is reestablished.
         """
-        pass
+
+        if index * 2 + 1 < len(self.store):
+            if index * 2 + 2 < len(self.store):
+                if self.store[index * 2 + 1].key > self.store[index * 2 + 2].key:
+                    if self.store[index * 2 + 1].key < self.store[index].key:
+                        self.swap(index, index * 2 + 1)
+                        self.heap_down(index * 2 + 1)
+                    else:
+                        return
+                else:
+                    if self.store[index * 2 + 2].key < self.store[index].key:
+                        self.swap(index, index * 2 + 2)
+                        self.heap_down(index * 2 + 2)
+                    else:
+                        return
+            else:
+                if self.store[index * 2 + 1].key < self.store[index].key:
+                    self.swap(index, index * 2 + 1)
+                    self.heap_down(index * 2 + 1)
```

Comment:
👀 Notice that the branches that do anything all culminate in swapping with an index and the making the recursive heap call on that other index.

Consider restructuring so that we calculate what the other index is, and then we could write the swap and heap calls just once.